File: dynamic/minPathSum.js
// very similar to grid traveler tabulation

var minPathSum = function (grid) {
  if (grid === null || grid.length === 0) {
    return 0;
  }
  const dp = new Array(grid.length)
    .fill()
    .map(() => Array(grid[0].length).fill(0));
  for (let i = 0; i < dp.length; i++) {
    for (let j = 0; j < dp[i].length; j++) {
      dp[i][j] += grid[i][j];
      if (i > 0 && j > 0) {
        // check whether above or left is smaller some
        dp[i][j] += Math.min(dp[i - 1][j], dp[i][j - 1]);
        // top row cant add dp[i][j - 1] if j is 0
      } else if (i > 0) {
        dp[i][j] += dp[i - 1][j];
        // left column cant add dp[i - 1][j] if i is 0
      } else if (j > 0) {
        dp[i][j] += dp[i][j - 1];
      }
    }
  }
  return dp[grid.length - 1][grid[0].length - 1];
};

File: dynamic/min_path_sum.py
grid = [[1,3,1],[1,5,1],[4,2,1]]
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        if len(grid) < 1: 
            return 0
        rows = len(grid)
        cols = len(grid[0])
        
        # add left colum
        for i in range(1, rows):
            grid[i][0] += grid[i - 1][0]
        
        # add top row
        for j in range(1, cols):
            grid[0][j] += grid[0][j - 1]
                
        for i in range(1, rows):
            for j in range(1, cols):
                grid[i][j] += min(grid[i - 1][j], grid[i][j - 1])
                
        return grid[rows - 1][cols - 1]




